Issue/7579 featured image confirmation

nbradbury opened the pull request:
Fixes #7579 - This PR is a refactoring of the device photo picker, which previously treated tapping a photo as an immediate selection. For example, if the user was choosing a featured image, simply tapping a photo would immediately upload it and set it as the featured image. This was inconsistent with how selection works in the new stock photo picker, plus it was simply annoying because users shouldn't expect a simple tap to result in uploading, etc.

So, this PR adds a confirmation step (via the CAB) so the user has a chance to approve the selection before anything gets uploaded.

To test, open a post in the editor and set a featured image, or tap the gravator on the Me page to select a new gravatar image.

![single photo picker](https://user-images.githubusercontent.com/3903757/38275789-07dbfd4c-3761-11e8-8f0f-4878180d98bd.gif)

hypest commented:
Did a quick test and it seems that the selection is lost on rotation @nbradbury .

Also, this is more weird, it seems that the list of thumbnails doesn't scroll fully in landscape on my Nexus 5X and the bottom images get overshadowed by the bottom bar.

hypest commented:
Thanks for the single-selection mode treatment @nbradbury . An issue I noticed is that after selecting a thumbnail and canceling the mode (say, tap "Back"), the thumbnail still shows as selected.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/photopicker/PhotoPickerAdapter.java:
```diff
@@ -213,49 +218,63 @@ private boolean isValidPosition(int position) {
+            int selectedIndex = mSelectedItems.indexOf(position);
+            if (selectedIndex > -1) {
+                mSelectedItems.remove(selectedIndex);
+            }
```
Can we simplify the search+removal into just `mSelectedItems.remove((Integer) position)` perhaps?

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/photopicker/PhotoPickerAdapter.java:
```diff
@@ -213,49 +218,63 @@ private boolean isValidPosition(int position) {
-        AniUtils.startAnimation(holder.mTxtSelectionCount, R.anim.pop);
+        ThumbnailViewHolder holder = getViewHolderAtPosition(position);
+        if (holder != null) {
+            if (canMultiselect()) {
```
The logic to update the `mTxtSelectionCount` here looks almost identical [to what happens in onBindViewHolder](https://github.com/wordpress-mobile/WordPress-Android/blob/0055e8fdf62265ec8e7d465daf98030cf191d96e/WordPress/src/main/java/org/wordpress/android/ui/photopicker/PhotoPickerAdapter.java#L168-L177). Can we perhaps unify it and extract it into a new method?

nbradbury commented:
@hypest I believe I've addressed all the issues you raised above, ready for another round!

nbradbury commented:
> It seems that on rotation the selected item is preserved, but the 'selected' mode is not: the "checkmark" CAB action is lost and the bottom bar is still visible.

Fixed in 312eb28.

> when a thumbnail gets deselected, its bubble icon animates OK but the icon itself is depicted as deselected

Fixed in 57b4397.

nbradbury commented:
> On the topic of rotation, not sure if this should be in scope for this PR but, the (bottom sheet) media picker seems to lose its selected items when trying to add media to a post body and rotate the device.

I had noticed that while testing, too. Let's make that a separate issue, since it existed prior to this PR.

hypest commented:
Thanks for the updates @nbradbury , :shipit: !
